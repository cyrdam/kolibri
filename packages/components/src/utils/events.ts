// TODO: Should be synchronized with enums/events.ts
export type KoliBriEventType =
	| 'blur'
	| 'change'
	| 'changePage'
	| 'changePageSize'
	| 'click'
	| 'close'
	| 'focus'
	| 'input'
	| 'mousedown'
	| 'reset'
	| 'selection-change' // todo remove
	| 'selectionChange'
	| 'sort'
	| 'submit'
	| 'toggle';

export function stopPropagation(event: Event): void {
	event.stopImmediatePropagation();
	event.stopPropagation();
}

function createKoliBriEvent<T>(type: KoliBriEventType, detail: T): CustomEvent {
	const event = new CustomEvent(`kol-${type}`, {
		bubbles: true,
		cancelable: true,
		composed: true,
		detail: detail,
	});
	return event;
}

function dispatchKoliBriEvent<T>(target: EventTarget, type: KoliBriEventType, detail?: T): boolean {
	const dispatch = target.dispatchEvent(createKoliBriEvent(type, detail));
	return dispatch;
}

/**
 * @deprecated
 */
export function tryToDispatchKoliBriEvent<T>(type: KoliBriEventType, target?: EventTarget, detail?: T): void {
	target && dispatchKoliBriEvent(target, type, detail);
}

export function dispatchDomEvent(target: HTMLElement, type: KoliBriEventType) {
	target.dispatchEvent(new Event(type, { bubbles: true, composed: true }));
}

export function preventDefaultAndStopPropagation(event: Event) {
	event.preventDefault();
	stopPropagation(event);
}
