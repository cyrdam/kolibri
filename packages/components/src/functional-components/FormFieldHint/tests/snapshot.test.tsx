import { h } from '@stencil/core';
import KolFormFieldHintFc from '../FormFieldHint';
import { renderFunctionalComponentToSpecPage } from '../../../utils/testing';

describe('KolFormFieldHintFc', () => {
	it('should render with hint', async () => {
		const hint = 'This is a hint';
		const id = 'test-id';
		const classNames = 'custom-class';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldHintFc hint={hint} id={id} class={classNames} />);

		expect(page.root).toMatchSnapshot();
		expect(page.root?.textContent).toContain(hint);
	});

	it('should not render without hint', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldHintFc />);

		expect(page.root).toBeNull();
	});

	it('should set the correct id', async () => {
		const hint = 'This is a hint';
		const id = 'test-id';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldHintFc hint={hint} id={id} />);

		expect(page.root?.id).toBe(`${id}-hint`);
	});

	it('should set the correct class names', async () => {
		const hint = 'This is a hint';
		const classNames = 'custom-class';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldHintFc hint={hint} class={classNames} />);

		expect(page.root?.className).toContain('hint');
		expect(page.root?.className).toContain(classNames);
	});
});
