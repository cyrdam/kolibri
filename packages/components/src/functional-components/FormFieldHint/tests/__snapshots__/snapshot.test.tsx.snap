// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KolFormFieldHintFc should render with hint 1`] = `
<span class="custom-class kol-form-field__hint" id="test-id-hint">
  This is a hint
</span>
`;
